### DISCUSSION TITLE
Space between key and value?

### DISCUSSION THREAD

**Discussion by hchiba1** [pg-format/pg-formatter#13] [General]
**with space:**
```python
# NODES
101 :person name: Alice country: "United States"
102 :person :student name: Bob country: Japan

# EDGES
101 -- 102 :same_school :same_club since: 2012
101 -> 102 :likes since: 2015
```

**without space:**
```python
# NODES
101 :person name:Alice country:"United States"
102 :person :student name:Bob country:Japan

# EDGES
101 -- 102 :same_school :same_club since:2012
101 -> 102 :likes since:2015
```

The latter one (without space) seems more compact, easier for me to handwrite and visually inspect.
Thus, I would not require the space, making it optional.

**Comment by nichtich** [comment]
I also prefer to allow value to directly follow property name. However this has some implications on allowed characters in unquoted property names and/or values:

~~~json
{"id":"n","labels":[],"properties":{"a\"key":["some:value"]}}
~~~

Could be serialized with or without space, with or without quotation:

    n "a\"key": "some:value"  # ok
    n "a\"key":"some:value"   # ok
    n "a\"key": some:value    # not allowed because value can be misread as property (?)
    n a"key: some:value       # not allowed because value  can be misread as property (?)
    n "a\"key":some:value     # may be allowed but colon in unquted value may also be forbidden
    n a"key: "some:value"     # may be allowed because quotation is not first character
    n a"key:"some:value"      # may be allowed but difficult to parse 
    n a"key:some:value        # not allowed because key is read as "a\"key:some:"

A clean solution may be to disallow quotation marks in unquoted property keys and values and disallow colon in property values.

**Comment by hchiba1** [comment]
`UNQUOTED_CHAR` is introduced (05dd5ed), thereby disallowing `"` and `:` in property keys and values.
(related to #11)

**Reply by nichtich** [reply]
There is nothing wrong with `:` in property keys (same as in node identifiers), but it should not be allowed in values.

**Reply by nichtich** [reply]
After some experimenting in writing PG format, I think that `:` can be allowed in unquoted keys and values as well, especially to support plain URIs. That's how it is implemented in pgraph right now.
